fileattrs/python.attr

```diff
@@ -1,4 +1,28 @@
-%__python_provides	%{_rpmconfigdir}/pythondeps.sh --provides
-%__python_requires	%{_rpmconfigdir}/pythondeps.sh --requires
+%__python_provides() %{lua:
+    -- Match buildroot/payload paths of the form
+    --    /PATH/OF/BUILDROOT/usr/bin/pythonMAJOR.MINOR
+    -- generating a line of the form
+    --    python(abi) = MAJOR.MINOR
+    -- (Don't match against -config tools e.g. /usr/bin/python2.6-config)
+    local path = rpm.expand('%1')
+    if path:match('/usr/bin/python%d+%.%d+$') then
```

Comment:
I am pretty sure this will also match `/var/lib/testing/framewrok/usr/bin/python3.8`, but that's **consistent with the previous behavior**. I was thinking about:

```lua
local buildroot = rpm.expand('%buildroot')
if path:match(buildroot .. '/usr/bin/python%d+%.%d+$')
```

But `%buildroot` can contain special characters (pluses, dots, etc.) and unfortunately there is no `escape` method in Lua.

If we want to go that road, I can work on that in a followup PR, hacking up [my own escape thing](https://stackoverflow.com/questions/9790688/escaping-strings-for-gsub).

Reply:
Rpm ships with a [regex Lua module](https://rpm.org/user_doc/lua#rex-extension) for reasons like this. Of course it'd help if said module was somehow documented (/me spots the TODO greetings from self from several years ago...)

Reply:
Documentation is for sissies, but is there an interactive shell for this? I am using `lua` but it doesn't have all the things RPM Lua has.

Reply:
As a matter of fact, there is. It's not for sissies either as there are not command history or even line editing available, but there is: `rpm --eval "%{lua:rpm.interactive()}"`

I personally find running `rpm --eval %{lua:...whatever...}` from the shell less painful.

Reply:
Wo needs line editing, right :laughing:

Reply:
Before I got that way, does `%{1}` even contain the buildroot?

Reply:
Because in Fedora's `python-rpm-generators`, we apparently have:

```
%__python_path ^((/usr/lib(64)?/python[[:digit:]]\\.[[:digit:]]+/.*\\.(py[oc]?|so))|(^%{_bindir}/python[[:digit:]]\\.[[:digit:]]+))$
```

And if that works, this entire problem is sovled on a higher level.

Reply:
It seems it does. Will try to upstream that change later.

Reply:
So in Fedora, we now have this:

https://src.fedoraproject.org/rpms/python-rpm-generators/pull-request/10

```
%__python_path ^((%{_prefix}/lib(64)?/python[[:digit:]]+\\.[[:digit:]]+/.*\\.(py[oc]?|so))|(%{_bindir}/python[[:digit:]]+\\.[[:digit:]]+))$
```

Is this applicable to upstream, or shall be keep our own regex?

Reply:
I don't see any problems with that.

BTW as for the %{buildroot} part: the classifier regex (`%__python_path` here) is applied on a path without %{buildroot} so the ^ is fine. However the %{1} argument *does* include the %{buildroot} - the situation is the same with "traditional" generators. This point probably deserves a clear note in the generator docs.

It also does make me wonder if this should be configurable somehow: many/most generators will look at the file contents, in which case they do need the buildroot prepended, but there *are* generators operate on the path alone, and in those cases it'd be somewhat easier to have the path as packaged without the buildroot. Such as here.

Reply:
Will open another PR.

> the classifier regex is applied on a path without %{buildroot} so the ^ is fine

I've verified this (as the `^` wuold obviously brek it otherwise).

>However the %{1} argument does include the %{buildroot}

That's why the `^` is only applied here and not in the generator code. I think it works well like this.

> This point probably deserves a clear note in the generator docs.

Indeed, it would be helpful if this is a documented behavior and not just implementation detail.

> It also does make me wonder if this should be configurable somehow: many/most generators will look at the file contents, in which case they do need the buildroot prepended, but there are generators operate on the path alone, and in those cases it'd be somewhat easier to have the path as packaged without the buildroot. Such as here.

The current behavior is fine. You can safely ignore the beginning of the path if you don't need it - as does this generator.

Reply:
https://github.com/rpm-software-management/rpm/pull/1179

Reply:
In the meanwhile:
https://github.com/rpm-software-management/rpm-web/commit/b34c9a9c5db052acf64334913bd87e73e6e8ef79

(but that doc is in desperate need of a full overhaul...)